File: runners/ruby/verify_fapi.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# Ruby reference runner for the FAPI 2.0 Message Signing corpus (fapi_messagesigning_v0).
#
# Independently reproduces every verdict in the frozen corpus. The RFC 9421
# Section 2.5 signing base is hand-assembled here; the PROFILE rules (mandated
# covered-component completeness, Content-Digest (RFC 9530) body binding,
# PS256/ES256 algorithm restriction) and the PS256/ES256 crypto are re-derived
# from the corpus `policy` block read at RUNTIME, NOT imported from the
# generator's oracle, so a runner passing is genuine agreement rather than an
# echo. Mirrors the Python reference runner (runners/python/verify_fapi.py)
# case for case.
#
# Self-contained: Ruby standard library only (json, base64, openssl, digest).
# RSA-PSS-SHA256 (PS256) and ECDSA P-256 (ES256, strict low-s) via OpenSSL; the
# EC public key is rebuilt from the uncompressed point and the raw r||s signature
# is re-encoded to DER for verification.
#
# Exit 0 iff every case matches. Corpus path: ARGV[0], else the repo corpus
# relative to this source file's directory (runners/ruby).
#
# Run:  ruby runners/ruby/verify_fapi.rb [corpus.json]

require "json"
require "base64"
require "openssl"
require "digest"

DEFAULT_PATH = "../../corpus/fapi_messagesigning_v0/fapi_messagesigning_v0.json"

class BuildError < StandardError; end

# ---- Section 1: RFC 9421 Section 2.5 signing base, by direct concatenation ----
# Raises BuildError when a covered component's value is missing (build impossible).
def fapi_signing_base(src, signature_params_raw)
  headers = {}
  (src["headers"] || {}).each { |k, v| headers[k.downcase] = v }

  lines = []
  (src["covered_components"] || []).each do |comp|
    name = comp.downcase
    val =
      case name
      when "@method"     then src["method"]
      when "@target-uri" then src["target_uri"]
      when "@status"     then src["status"].nil? ? nil : src["status"].to_s
      when "@authority"  then src["authority"]
      when "@path"       then src["path"]
      else headers[name]
      end
    raise BuildError, "missing covered component: #{name}" if val.nil?
    lines << "\"#{name}\": #{val}"
  end
  lines << "\"@signature-params\": #{signature_params_raw}"
  lines.join("\n")
end

# ---- Section 2: required covered-component completeness ----
def required_coverage_ok(message_type, covered, policy)
  key = message_type == "request" ? "required_covered_request" : "required_covered_response"
  have = (covered || []).map(&:downcase)
  policy[key].all? { |r| have.include?(r.downcase) }
end

# ---- Section 3: Content-Digest (RFC 9530) body binding ----
def content_digest_ok(body_b64, header_value, covered, policy)
  return false unless (covered || []).map(&:downcase).include?("content-digest")
  algorithm, rest = header_value.split("=", 2)
  return false if rest.nil?
  algorithm = algorithm.strip.downcase
  return false unless policy["content_digest_algorithms"].include?(algorithm)
  body = Base64.decode64(body_b64)
  digest = algorithm == "sha-256" ? Digest::SHA256.digest(body) : Digest::SHA512.digest(body)
  expect = "#{algorithm}=:#{Base64.strict_encode64(digest)}:"
  header_value.strip == expect
end

# ---- Section 5: PS256 (RSA-PSS SHA-256, salt 32) over the signing base ----
def ps256_ok(base, sig, spki_der)
  key = OpenSSL::PKey.read(spki_der)
  key.verify_pss("SHA256", sig, base, salt_length: 32, mgf1_hash: "SHA256")
rescue StandardError
  false
end

# ---- Section 6: ES256 (ECDSA P-256 SHA-256, strict low-s) over the signing base ----
def es256_ok(base, sig_raw, pub_uncompressed, require_low_s, policy)
  return false unless sig_raw.bytesize == 64
  r = sig_raw[0, 32].unpack1("H*").to_i(16)
  s = sig_raw[32, 32].unpack1("H*").to_i(16)
  return false if require_low_s && s > policy["p256_n"].to_i / 2

  group = OpenSSL::PKey::EC::Group.new("prime256v1")
  # Point.new validates the point is on the curve (raises otherwise)
  OpenSSL::PKey::EC::Point.new(group, OpenSSL::BN.new(pub_uncompressed.unpack1("H*"), 16))
  spki = OpenSSL::ASN1::Sequence([
    OpenSSL::ASN1::Sequence([
      OpenSSL::ASN1::ObjectId("id-ecPublicKey"),
      OpenSSL::ASN1::ObjectId("prime256v1")
    ]),
    OpenSSL::ASN1::BitString(pub_uncompressed)
  ])
  key = OpenSSL::PKey.read(spki.to_der)
  der = OpenSSL::ASN1::Sequence([OpenSSL::ASN1::Integer(r), OpenSSL::ASN1::Integer(s)]).to_der
  key.verify(OpenSSL::Digest.new("SHA256"), der, base)
rescue StandardError
  false
end

# ---- driver ----
def hexb(s) = [s].pack("H*")

path = ARGV[0] || DEFAULT_PATH
corpus = JSON.parse(File.read(path))
policy = corpus["policy"]
results = []

record = lambda { |section, note, ok| results << [section, note, ok] }

# 1. fapi_signing_base
corpus["fapi_signing_base"].each do |c|
  want = Base64.decode64(c["signing_base_b64"]).force_encoding("UTF-8")
  ok =
    begin
      # build FIRST; `c["ok"] && build(...)` would short-circuit negative cases
      # and never exercise the build-failure path.
      built = fapi_signing_base(c["in"], c["signature_params_raw"])
      c["ok"] && built == want
    rescue BuildError
      !c["ok"]
    end
  record.call("fapi_signing_base", c["note"], ok)
end

# 2. fapi_required_coverage
corpus["fapi_required_coverage"].each do |c|
  record.call("fapi_required_coverage", c["note"],
              required_coverage_ok(c["message_type"], c["covered_components"], policy) == c["expect_accept"])
end

# 3. fapi_content_digest
corpus["fapi_content_digest"].each do |c|
  record.call("fapi_content_digest", c["note"],
              content_digest_ok(c["body_b64"], c["content_digest"], c["covered_components"], policy) == c["expect_accept"])
end

# 4. fapi_alg
corpus["fapi_alg"].each do |c|
  record.call("fapi_alg", c["note"],
              policy["allowed_algs"].include?(c["alg"].downcase) == c["expect_accept"])
end

# 5. fapi_ps256_verify
corpus["fapi_ps256_verify"].each do |c|
  base = Base64.decode64(c["signing_base_b64"])
  valid = ps256_ok(base, hexb(c["sig_hex"]), hexb(c["pub_spki_hex"]))
  record.call("fapi_ps256_verify", c["note"], valid == c["expect_valid"])
end

# 6. fapi_es256_verify
corpus["fapi_es256_verify"].each do |c|
  base = Base64.decode64(c["signing_base_b64"])
  valid = es256_ok(base, hexb(c["sig_raw_hex"]), hexb(c["pub_uncompressed_hex"]),
                   c["require_low_s"] == true, policy)
  record.call("fapi_es256_verify", c["note"], valid == c["expect_valid"])
end

fails = results.reject { |_, _, ok| ok }
fails.each { |section, note, _| puts "FAIL  [#{section}] #{note}" }
total = results.length
puts "\nruby (fapi): #{total - fails.length}/#{total} cases matched"
exit(fails.empty? ? 0 : 1)

File: runners/elixir/verify_fapi.exs
# Elixir runner for the FAPI 2.0 Message Signing corpus (fapi_messagesigning_v0).
#
# Independent Elixir port of the Python reference runner
# (runners/python/verify_fapi.py). Every PROFILE rule (RFC 9421 Section 2.5
# signing base, mandated covered-component completeness, Content-Digest
# (RFC 9530) body binding, PS256/ES256 algorithm restriction) and the
# PS256/ES256 signature verdicts are re-implemented here directly from the
# corpus `policy` block, read at RUNTIME, NOT imported from a shared oracle, so
# a runner passing is genuine agreement rather than an echo. This mirrors the
# Python reference runner case for case.
#
# Self-contained, matching the sibling negative_v1.exs / verify_wba.exs: JSON via
# jason, RSA-PSS-SHA256 and ECDSA-P256-SHA256 via Erlang :public_key / :crypto.
#
# Corpus path: argv[0], else the repo corpus relative to this source directory.
#
# Run:  cd runners/elixir && elixir verify_fapi.exs [corpus.json]

Mix.install([{:jason, "~> 1.4"}])

import Bitwise

defmodule SigningBaseError, do: (defexception message: "")

defmodule VerifyFapi do
  @default_path "../../corpus/fapi_messagesigning_v0/fapi_messagesigning_v0.json"

  # ------------------------------------------------------------------
  # Rule 1: RFC 9421 Section 2.5 signing base, hand-built by direct
  # concatenation (mode "rfc9421": derived components from the `in` fields,
  # every other covered name from headers, then the @signature-params line).
  # Raises SigningBaseError when a covered component's value is missing
  # (build impossible); the driver maps that to "unbuildable".
  # ------------------------------------------------------------------
  def build_signing_base(input, params_raw) do
    headers = for {k, v} <- input["headers"] || %{}, into: %{}, do: {String.downcase(k), v}

    lines =
      Enum.map(input["covered_components"], fn comp ->
        c = String.downcase(comp)

        value =
          case c do
            "@method" -> fetch!(input, "method", "@method covered but method not supplied")
            "@target-uri" -> fetch!(input, "target_uri", "@target-uri covered but target_uri not supplied")
            "@authority" -> String.downcase(fetch!(input, "authority", "@authority covered but authority not supplied"))
            "@path" -> fetch!(input, "path", "@path covered but path not supplied")
            "@scheme" -> String.downcase(fetch!(input, "scheme", "@scheme covered but scheme not supplied"))
            "@status" ->
              case input["status"] do
                nil -> raise(SigningBaseError, message: "@status covered but status not supplied")
                s -> to_string(s)
              end

            _ ->
              if String.starts_with?(c, "@"),
                do: raise(SigningBaseError, message: "unsupported derived component: #{comp}")

              case Map.fetch(headers, c) do
                {:ok, hv} when not is_nil(hv) -> hv
                _ -> raise(SigningBaseError, message: "covered header #{comp} not present")
              end
          end

        "\"#{c}\": #{value}"
      end)

    Enum.join(lines ++ ["\"@signature-params\": #{params_raw}"], "\n")
  end

  defp fetch!(map, key, msg) do
    case Map.fetch(map, key) do
      {:ok, v} when not is_nil(v) -> v
      _ -> raise(SigningBaseError, message: msg)
    end
  end

  # ------------------------------------------------------------------
  # Rule 2: mandated covered-component completeness. The required set is keyed
  # by message_type; the required names (lowercased) must be a subset of the
  # covered names (lowercased).
  # ------------------------------------------------------------------
  def coverage_ok(message_type, covered, policy) do
    key = if message_type == "request", do: "required_covered_request", else: "required_covered_response"
    have = MapSet.new(covered || [], &String.downcase/1)
    Enum.all?(policy[key], fn r -> MapSet.member?(have, String.downcase(r)) end)
  end

  # ------------------------------------------------------------------
  # Rule 3: Content-Digest (RFC 9530) body binding. content-digest must be
  # covered; the header's algorithm token (lowercased) must be an allowed one;
  # and the header must equal "<alg>=:<base64(hash(body))>:" exactly.
  # ------------------------------------------------------------------
  def content_digest_ok(body_b64, header_value, covered, policy) do
    covered_lc = MapSet.new(covered || [], &String.downcase/1)

    cond do
      not MapSet.member?(covered_lc, "content-digest") -> false
      true ->
        case String.split(header_value, "=", parts: 2) do
          [alg_raw, _] ->
            algorithm = alg_raw |> String.trim() |> String.downcase()

            cond do
              algorithm not in policy["content_digest_algorithms"] -> false
              true ->
                body = Base.decode64!(body_b64)
                hash_alg = if algorithm == "sha-256", do: :sha256, else: :sha512
                digest = :crypto.hash(hash_alg, body)
                expect = "#{algorithm}=:#{Base.encode64(digest)}:"
                String.trim(header_value) == expect
            end

          _ -> false
        end
    end
  end

  # ------------------------------------------------------------------
  # Rule 5: PS256 (RSA-PSS SHA-256, salt length 32) over the raw signing-base
  # bytes, with the RSA public key extracted from its SubjectPublicKeyInfo DER.
  # Mirrors the sibling negative_v1.exs RSA key parse.
  # ------------------------------------------------------------------
  def ps256_ok(base, sig, spki_der) do
    try do
      spki_info = :public_key.der_decode(:SubjectPublicKeyInfo, spki_der)
      {:RSAPublicKey, n, e} = :public_key.der_decode(:RSAPublicKey, elem(spki_info, 2))

      :crypto.verify(:rsa, :sha256, base, sig, [e, n],
        [{:rsa_padding, :rsa_pkcs1_pss_padding}, {:rsa_pss_saltlen, 32}, {:rsa_mgf1_md, :sha256}])
    rescue
      _ -> false
    end
  end

  # ------------------------------------------------------------------
  # Rule 6: ES256 (ECDSA P-256 SHA-256). The signature is raw 64-byte r||s.
  # Reject high-s (s > n/2) when low-s is required, else convert r||s to a DER
  # ECDSA signature (SEQUENCE of two INTEGERs) and verify against the
  # uncompressed public point.
  # ------------------------------------------------------------------
  def es256_ok(base, sig_raw, pub_uncompressed, require_low_s, policy) do
    if byte_size(sig_raw) != 64 do
      false
    else
      r = :binary.decode_unsigned(binary_part(sig_raw, 0, 32))
      s = :binary.decode_unsigned(binary_part(sig_raw, 32, 32))
      n = String.to_integer(policy["p256_n"])

      cond do
        require_low_s and s > div(n, 2) -> false
        true ->
          der = <<0x30, byte_size(der_int(r) <> der_int(s))>> <> der_int(r) <> der_int(s)

          try do
            :crypto.verify(:ecdsa, :sha256, base, der, [pub_uncompressed, :secp256r1])
          rescue
            _ -> false
          end
      end
    end
  end

  defp der_int(i) do
    bytes = :binary.encode_unsigned(i)
    bytes = if (:binary.first(bytes) &&& 0x80) != 0, do: <<0>> <> bytes, else: bytes
    <<0x02, byte_size(bytes)>> <> bytes
  end

  # ------------------------------------------------------------------
  # Driver.
  # ------------------------------------------------------------------
  defp hexb(s), do: Base.decode16!(s, case: :lower)

  def main(argv) do
    path = List.first(argv) || @default_path
    corpus = Jason.decode!(File.read!(path))
    policy = corpus["policy"]
    results = []

    results =
      results ++
        Enum.map(corpus["fapi_signing_base"], fn c ->
          want = Base.decode64!(c["signing_base_b64"])
          # Build FIRST; `c["ok"] and build(...)` would short-circuit negative
          # cases and never exercise the build-failure path.
          ok =
            try do
              built = build_signing_base(c["in"], c["signature_params_raw"])
              c["ok"] and built == want
            rescue
              _ -> not c["ok"]
            end

          {ok, "fapi_signing_base", c["note"]}
        end)

    results =
      results ++
        Enum.map(corpus["fapi_required_coverage"], fn c ->
          {coverage_ok(c["message_type"], c["covered_components"], policy) == c["expect_accept"],
           "fapi_required_coverage", c["note"]}
        end)

    results =
      results ++
        Enum.map(corpus["fapi_content_digest"], fn c ->
          {content_digest_ok(c["body_b64"], c["content_digest"], c["covered_components"], policy) ==
             c["expect_accept"], "fapi_content_digest", c["note"]}
        end)

    results =
      results ++
        Enum.map(corpus["fapi_alg"], fn c ->
          {(String.downcase(c["alg"]) in policy["allowed_algs"]) == c["expect_accept"], "fapi_alg",
           c["note"]}
        end)

    results =
      results ++
        Enum.map(corpus["fapi_ps256_verify"], fn c ->
          base = Base.decode64!(c["signing_base_b64"])
          ok = ps256_ok(base, hexb(c["sig_hex"]), hexb(c["pub_spki_hex"]))
          {ok == c["expect_valid"], "fapi_ps256_verify", c["note"]}
        end)

    results =
      results ++
        Enum.map(corpus["fapi_es256_verify"], fn c ->
          base = Base.decode64!(c["signing_base_b64"])
          ok =
            es256_ok(base, hexb(c["sig_raw_hex"]), hexb(c["pub_uncompressed_hex"]),
              c["require_low_s"] == true, policy)

          {ok == c["expect_valid"], "fapi_es256_verify", c["note"]}
        end)

    fails = Enum.filter(results, fn {ok, _, _} -> not ok end)
    Enum.each(fails, fn {_, s, n} -> IO.puts("FAIL  [#{s}] #{n}") end)
    matched = Enum.count(results, fn {ok, _, _} -> ok end)
    total = length(results)
    IO.puts("\nelixir (fapi): #{matched}/#{total} cases matched")
    System.halt(if length(fails) == 0, do: 0, else: 1)
  end
end

VerifyFapi.main(System.argv())
